### PR TITLE
assertion operations

### PR DESCRIPTION
**Pull request by yura-ata**

### CODE REVIEW

File: src/main/scala/org/scanet/core/CoreOp.scala [hunk 1]
```diff
@@ -65,19 +65,43 @@ import scala.language.higherKinds
...
    */
   def cast[A: TensorType, B: TensorType](op: F[A]): F[B]
+
+  /** Return current op as leaf operation - means it doesnt produce output
+   * and can be evaluated in graph only if added as dependant operation.
+   *
+   * @see dependsOn
+   * @return current leaf node
+   */
+  def asLeaf[A: TensorType](op: F[A]): F[Nothing]
```
Comment (pashashiz) [inline src/main/scala/org/scanet/core/CoreOp.scala]: Not sure that `asLeaf` is the best name, basically in this expression `c = a + b` both `a` and `b` are leaves. Maybe we can call it something like `asUnit`, `asVoid`, `noop`, ..?

File: src/test/scala/org/scanet/core/CoreOpsSpec.scala [hunk 2]
```diff
@@ -60,4 +60,40 @@ class CoreOpsSpec extends AnyFlatSpec with Matchers {
...
+  }
+
+  "assert that" should "should fail when dependent condition is false" in {
+    the [IllegalArgumentException] thrownBy {
+      val a = Tensor.vector(1, 2).const
+      val b = 0.const
+      val c = (a div b) << assertThat(b gt 0.const, b)
+
+      c.eval should be(Tensor.scalar(3))
+    } should have message "assertion failed: [0]\n\t [[{{node Assert_0}}]]"
```
Comment (pashashiz) [inline src/test/scala/org/scanet/core/CoreOpsSpec.scala]: I am wondering if there is a way to have a custom assertion message, I guess that might be impossible though...

Comment (pashashiz) [reply src/test/scala/org/scanet/core/CoreOpsSpec.scala]: What if we add `print error` dependency under the hood so we could also print a nice assertion error?

File: src/main/scala/org/scanet/core/CoreOp.scala (hunk 1 above)
Comment (yura-ata) [reply src/main/scala/org/scanet/core/CoreOp.scala]: renamed into `asVoid`

File: src/test/scala/org/scanet/core/CoreOpsSpec.scala (hunk 2 above)
Comment (yura-ata) [reply src/test/scala/org/scanet/core/CoreOpsSpec.scala]: added message template for assertion erros (had to move assert from core into string ops to have access to `format` op - not to add circular dep into core ops)

**Comment by pashashiz** [comment]
Awesome!

File: src/test/scala/org/scanet/strings/StringOpsTest.scala [hunk 3]
```diff
@@ -107,4 +107,40 @@ class StringOpsTest extends AnyFlatSpec with Matchers {
...
+    val c = (a plus b) << assertThat((a gt 0.const) and (b gt 0.const), "{} {}", a, b)
+
+    c.eval should be(Tensor.scalar(3))
+  }
+
+  "assert that" should "should fail when dependent condition is false" in {
+    the [IllegalArgumentException] thrownBy {
+      val a = Tensor.vector(1, 2).const
+      val b = 0.const
+      val c = (a div b) << assertThat(b gt 0.const, "b was {}, should be > 0", b)
```
Comment (pashashiz) [inline src/test/scala/org/scanet/strings/StringOpsTest.scala:141]: Would be nice if you could check with alias, like:
```
val c = (a div b) << assertThat(b gt 0.const, "b was {}, should be > 0", b).as("b_is_positive_assertion")
```

**Pull request merged**